### PR TITLE
Add `backup_log` system table

### LINKED ISSUE #53221: Add a new backup_log table
**Use case**

@e-mars in [this](https://github.com/ClickHouse/ClickHouse/issues/43995) issue put forward an idea of `system.backups` table persistence (and I'm already working on its implementation).

@mlex in the same issue suggested that we might as well add a new `backup_log` table as an _alternative_ to making `system.backups` table persistent.

@UnamedRus please also take a look at this issue.

**Describe the solution you'd like**

My suggestion is to introduce `backup_log` table _along with_ the persistent `system.backups` table (which is yet to be implemented).

This table would apparently be one one of the "system logs" along with `query_log`, `session_log` etc. Those system logs can be conveniently turned on and off in the config file, their engine (e.g. `MergeTree`) and other properties are also configurable and the implementation is next to trivial. Moreover, @chhetripradeep even suggested an adequate solution (yet without a pull request).

So, if a new `backup_log` table would come in handy, I'm ready to implement it and create a pull request if necessary.

Kindly looking forward to any feedback from the community.

### PR DESCRIPTION
**Pull request by sirvickr**
Closes #53221

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a new system log table `backup_log` to track all `BACKUP` and `RESTORE` operations 

### Documentation entry for user-facing changes
- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by sirvickr** [comment]
@vitlibar could you please take a look at this pull request?

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit a7813260c0914b36750e6850d20fc45117bd1c44 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53638/a7813260c0914b36750e6850d20fc45117bd1c44/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by sirvickr** [comment]
Hello guys. I'm trying to figure out what's wrong with the CI jobs. Have they been cancelled manually or disrupted by some preceding failure?

**Comment by chhetripradeep** [comment]
You can try add an empty commit to re-trigger the build.

**Comment by sirvickr** [comment]
Sure, thanks. I might as well merge master into my branch and thus re-launch the CI. 
I was just wondering why the jobs are massively cancelled at times.

**Comment by rvasin** [comment]
> Sure, thanks. I might as well merge master into my branch and thus re-launch the CI. I was just wondering why the jobs are massively cancelled at times.

Maybe it's some kind of a bug in CI. I hope someone from ClickHouse like @vitlibar will explain it.

**Comment by rvasin** [comment]
@vitlibar, could you please review this PR? Because @sirvickr is already working on the related task about `system.backups` persistence (see [issue](https://github.com/ClickHouse/ClickHouse/issues/43995)) and his new code is based on current PR. 

So ideally it would be nice to merge the current PR to master at the end of this week. Then @sirvickr could create new PR about the system.backups persistence at the beginning or the middle of the next week.

File: src/Backups/BackupsWorker.cpp [hunk 1]
```diff
@@ -44,7 +45,7 @@ namespace ErrorCodes
     extern const int CONCURRENT_ACCESS_NOT_SUPPORTED;
 }
 
-using OperationID = BackupsWorker::OperationID;
+using OperationID = BackupOperationInfo::ID;
```
Comment (vitlibar) [inline src/Backups/BackupsWorker.cpp:48]: Maybe it's better to declare `BackupOperationID` instead of `BackupOperationInfo::ID` just to be shorter.

File: src/Backups/BackupsWorker.cpp [hunk 2]
```diff
@@ -449,9 +451,10 @@ void BackupsWorker::doBackup(
         backup.reset();
 
         LOG_INFO(log, "{} {} was created successfully", (backup_settings.internal ? "Internal backup" : "Backup"), backup_name_for_logging);
-        setStatus(backup_id, BackupStatus::BACKUP_CREATED);
         /// NOTE: we need to update metadata again after backup->finalizeWriting(), because backup metadata is written there.
         setNumFilesAndSize(backup_id, num_files, total_size, num_entries, uncompressed_size, compressed_size, 0, 0);
```
Comment (vitlibar) [inline src/Backups/BackupsWorker.cpp:455]: setNumFilesAndSize() doesn't update the backup log, does it? It's better to have a comment about that.

File: tests/integration/test_backup_log/test.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+@pytest.fixture(scope="module", autouse=True)
+def start_cluster():
+    try:
+        cluster.start()
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def back_table_up(backup_name):
```
Comment (vitlibar) [inline tests/integration/test_backup_log/test.py:24]: backup_table() sounds better

File: tests/integration/test_backup_log/test.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+from helpers.cluster import ClickHouseCluster
+from helpers.test_tools import TSV
+
+
+cluster = ClickHouseCluster(__file__)
+
+instance = cluster.add_instance(
+    "instance",
+    main_configs=["configs/config.xml", "configs/config.d/backups.xml"],
+    stay_alive=True,
```
Comment (vitlibar) [inline tests/integration/test_backup_log/test.py:11]: `stay_alive` is not necessary, your test doesn't restart the server

Comment (sirvickr) [reply tests/integration/test_backup_log/test.py:11]: Actually, it restarts the server (`instance.restart_clickhouse()` on line 50), just to ensure that the `MergeTree`-based system log stays persistent afterwards. 
Without `stay_alive` parameter set to `True`, the `restart_clickhouse` method in `cluster.py` throws an exception `clickhouse can be stopped only with stay_alive=True instance`.

File: src/Backups/BackupsWorker.cpp (hunk 2 above)
Comment (sirvickr) [reply src/Backups/BackupsWorker.cpp:455]: Each backup/restore operation creates two entries in the `backup_log`:
1. When the operation starts, it records start `date` and `time (us)` of the log entry itself, as well ad `id`, `name`, initial `status` and `start_time` of the operation.
2. When the operation ends, it records the same fields plus `end_time` and either `num_files`, `total_size`, `num_entries`, `uncompressed_size`, `compressed_size`, `files_read`, `bytes_read` in case of success of `error` in case of failure.
The `setNumFilesAndSize()` method updates fields of `BackupOperationInfo` `num_files`, `total_size`, `num_entries`, `uncompressed_size`, `compressed_size`, `num_read_files` and `num_read_bytes` of current operation's entry in the `infos` map, which in turn is inserted as the second entry after the operation has finished.
In a nutshell, `setNumFilesAndSize()` updates the current operation's entry in the `infos` map, and `setStatus()` flushes it to the log as the second line.
Probably, in the future we will come up (maybe not) with some intermediate steps to reflect the progress of long operations.

File: tests/integration/test_backup_log/test.py (hunk 3 above)
Comment (sirvickr) [reply tests/integration/test_backup_log/test.py]: Done.

File: src/Backups/BackupsWorker.cpp (hunk 2 above)
Comment (vitlibar) [reply src/Backups/BackupsWorker.cpp:455]: I meant the function `setNumFilesAndSize()` could call

```
    if (backup_log)
        backup_log->add(BackupLogElement{info});
```

just like `setStatus()` does.

Comment (sirvickr) [reply src/Backups/BackupsWorker.cpp:455]: Sure thing, got it.
As I developed this feature, I also thought about adding a record to `backup_log` within `setNumFilesAndSize()` method, but eventually decided to do it immediately in a subsequent `setStatus()` call which appears to be inevitable regardless of the final operation status - failure or success.
Done (I placed the comment at the beginning of the `setNumFilesAndSize()` method).

**Comment by sirvickr** [comment]
@vitlibar, what do you think about this pull request? If you have any concerns or ideas, please feel free to share.

**Review by vitlibar — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@sirvickr a bug has been found automatically: https://s3.amazonaws.com/clickhouse-test-reports/35961/c9753d7fd99050f0b6d4c4c5a02ab2c265239879/stateless_tests__tsan__[3_5]/stderr.log

**Comment by sirvickr** [comment]
Thank you, I'll take a close look.

**Comment by sirvickr** [comment]
I think I found the origin of the bug. The data race is possible due to the assignment of the `backup_log` member in the `BackupsWorker::start()` method which is called every time the query is executed within an arbitrary thread.
One of the possible decisions is to protect `backup_log` with a mutex but the whole decision to initialize it in the `BackupsWorker::start()` method looks like a bad idea.
The better decision would be to move its initialization to the `BackupsWorker`'s constructor, just like this:
image
I'm going to double check everything locally and make a fixing PR ASAP.

**Comment by alexey-milovidov** [comment]
Thank you! Please do :pray:
